Feature/logged in riblet

randychoi95 opened the pull request:
# LoggedIn Riblet 생성 및 초기 화면 구성
- LoggedIn Riblet의 화면으로 UITabbarController를 상속받아 tabbar화면 구현
- AppRootInteractor에서 이전에 로그인하였는지 판별
  - 이전에 로그인 했을 경우: LoggedIn attach
  - 이전에 로그인 하지 않았을 경우: LoggedOut attach -> Login -> LoggedOut detach -> LoggedIn attach

# 다음 진행사항
- 사용자 정보 LoggedIn Riblet에 전달
- 하위 Riblet Attach

randychoi95 committed:
- LoggedIn Riblet 생성

randychoi95 committed:
:sparkles: LoggedIn Riblet 추가 및 화면 구성

- UITabbarController를 상속받아 tabbar 구현
- 첫 로그인 / 자동로그인 후 LoggedIn Riblet attach 구현

Commit:
:package: MyLocation Riblet 추가

- LoggedIn Riblet에서 MyLocation Riblet Attach
- Tabbar Item 크기 수정
- UINavigationController로 감싸기 추가

Commit:
:see_no_evil: .gitignore 내용 수정

f-lab-max commented on PlaceStory/AppRoot/Sources/AppRoot/AppRootViewController.swift:
```diff
@@ -34,4 +32,19 @@ final class AppRootViewController: UIViewController, AppRootPresentable, AppRoot
         present(viewController.uiviewController, animated: true, completion: nil)
     }
+    
+    func dismiss(viewController: ViewControllable) {
```
함수명은 vc를 dismiss 한다 인데, loggedin을 붙이는 작업까지 해주고 있습니다. 둘을 분리하거나 네이밍을 좀 더 정확하게 수정하는게 좋겠습니다.

randychoi95 replied:
vc를 dismiss하고 loggedIn 붙이게 하기 위해서 해당 작업을 하였습니다.
method 네이밍을 좀 더 정확하게 수정을 해보겠습니다.

f-lab-max commented on PlaceStory/AppRoot/Sources/AppRoot/AppRootViewController.swift:
```diff
@@ -34,4 +32,19 @@ final class AppRootViewController: UIViewController, AppRootPresentable, AppRoot
+    }
+    
+    func presentLoggedIn() {
+        DispatchQueue.main.async { [weak self] in
```
(1) 메인큐에 디스패치하는 코드가 왜 필요한가요? 필요하다고 한다면, 여기서 해주는게 좋은지, 아니면 호출하는 쪽에서 담당하는게 좋을지 고민해보시면 좋겠습니다. 

(2) 인터랙터가 바로 attach를 하지 않고 VC를 거쳐서 attachLoggedIn을 호출하는 이유는 무엇인가요?

randychoi95 replied:
![스크린샷 2023-12-19 오후 9 18 36](https://github.com/f-lab-edu/PlaceStory/assets/60861313/24213bc9-7c4c-4aa6-bd46-992b57668168)
위 이유로 디스패치 코드를 넣었습니다.
Interactor에서 위 메서드를 호출하는데 디스패치큐는 UIKit을 임포트해야하기 때문에 ViewController에서 사용하였습니다.
그래서 VC를 거쳐서 호출하게 되었습니다.

f-lab-max replied:
interactor에 UIKit을 안쓰려고 하는것은 매우 좋은 방향입니다. 다만 DispatchQueue는 UIKit이 아니라 Foundation에 있습니다.

애초에 왜 background thread에서 메서드가 호출되어서 저 에러가 발생했는지 근본적인 원인을 확인바랍니다. 그러고나서 main queue에 특정 메서드를 호출해야하는 책임을 누가 지는게 좋을지 비교 고민해보면 좋겠습니다.

randychoi95 replied:
Realm에서 사용자 정보를 읽어오는 작업이 백그라운드 스레드에서 이루어지고 있습니다. 이 정보를 바탕으로 UI를 attach하는데 UI 관련 작업은 메인 스레드에서 이루어져야 하므로 main queue에 특정 메서드를 호출하게 적용하였습니다.

Interactor에서 사용자 정보 읽어오는 로직을 처리하고 그 결과로 View를 attach하기 때문에, 이 경우에 Interactor에서 메인 큐를 사용하는 것이 적절하다고 판단해서 수정을 해보았습니다.

f-lab-max commented on PlaceStory/AppRoot/Sources/AppRoot/AppRootViewController.swift:
```diff
@@ -34,4 +32,19 @@ final class AppRootViewController: UIViewController, AppRootPresentable, AppRoot
     }
+    
+    func dismiss(viewController: ViewControllable) {
+        viewController.uiviewController.dismiss(animated: true) { [weak self] in
```
라우팅(present, dismiss, push 등)은 라우터의 역할입니다. VC는 뷰의 역할만을 하는것이 좋습니다.

randychoi95 replied:
수정해보겠습니다.

Commit:
:zap: 리뷰반영

- 라우팅을 라우터에서 수행하게 수정하고, VC에서 라우팅 관련 작업 삭제
- Realm에서 read하는 작업이 백그라운드 스레드에서 작업을 하고 있어 UI 관련 작업은 메인 스레드에서 실행해야 해서 DispatchQueue를 추가
- VC를 거치지 않고 바로 attach 할 수 있어, VC 부분 삭제 후 바로 attach 할 수 있게 수정

f-lab-max commented on PlaceStory/AppRoot/Sources/AppRoot/AppRootViewController.swift:
```diff
@@ -9,9 +9,7 @@ import ModernRIBs
-    // TODO: Declare properties and methods that the view controller can invoke to perform
-    // business logic, such as signIn(). This protocol is implemented by the corresponding
-    // interactor class.
+    func attachLoggedIn()
```
UI에서 "logged in 붙여라" 라고 알리는 것이 맞을까요? UI가 그런 결정을 내리는 곳인가 고민이 필요해보입니다. 인터페이스, 네이밍 관점에서 생각해보시길 바랍니다. UI의 역할이 무엇이며 interactor의 역할이 무엇인지, 그에 따른 각각의 인터페이스의 의미는 무엇일지 고민해보고 네이밍이 개선되면 좋겠습니다.

randychoi95 replied:
네 알겠습니다.
해당 부분은 기존 VC에서 사용할 때 남았던 부분이라 지금은 현재 사용하지 않아 삭제해도 되는 부분이라 삭제하겠습니다.
그럼에도 위의 부분에 대해서 한번 더 고민하면서 후에는 잘 적용해보겠습니다.

f-lab-max reviewed (commented):
PR을 나눌때 ribs 템플릿 추가하는 것과 구현 파트를 나눠서 올리면 더 좋을거 같습니다. 리뷰어 입장에서 PR을 볼때 무엇을 봐야할지 좀 더 집중할 수 있게 됩니다.

f-lab-max reviewed (commented):
핵심 로직이 추가되고 있는 지금부터 유닛테스트 작성도 같이 병행되면 좋겠습니다

randychoi95 committed:
- Listener의 "attachLoggedIn" 메소드 미사용으로 인한 삭제
